mpisppy/opt/sc.py

```diff
@@ -0,0 +1,121 @@
+from mpisppy.spbase import SPBase
+import pyomo.environ as pyo
+import logging
+import parapint
+from typing import List, Callable, Dict, Optional, Tuple, Any, Union
+from pyomo.core.base.block import _BlockData
+from pyomo.core.base.var import _GeneralVarData
+from mpi4py import MPI
+
+
+logger = logging.getLogger('mpisppy.sc')
+
+
+SCOptions = parapint.algorithms.IPOptions
+
+
+def _assert_continuous(m: _BlockData):
+    for v in m.component_data_objects(pyo.Var, descend_into=True, active=True):
+        if not v.is_continuous():
+            raise RuntimeError(f'Variable {v} in block {m} is not continuous; The Schur-Complement method only supports continuous problems.')
+
+
+def _get_active_objective(m: _BlockData):
+    res = None
+    for obj in m.component_data_objects(pyo.Objective, descend_into=True, active=True):
+        if res is not None:
+            raise RuntimeError(f'Found multiple active objectives on {m}')
+        res = obj
+    return res
+
+
+class _SCInterface(parapint.interfaces.MPIStochasticSchurComplementInteriorPointInterface):
+    def __init__(self,
+                 local_scenario_models: Dict[str, _BlockData],
+                 all_scenario_names: List[str],
+                 comm: MPI.Comm,
+                 ownership_map: Dict):
+        self.local_scenario_models = local_scenario_models
+
+        models = list(local_scenario_models.values())
+        ref_model = models[0]
+        models = models[1:]
+        self.nonant_vars = list()
+
+        for node in ref_model._mpisppy_node_list:
+            ndn = node.name
+            for ndx, v in enumerate(node.nonant_vardata_list):
+                self.nonant_vars.append((ndn, ndx))
```

Comment:
`self.nonant_vars = list(ref_model._mpisppy_data.nonant_indices.keys())` would accomplish the same thing.

Reply:
This is very helpful. Thanks.